### PR TITLE
WIP: [DO_NOT_MERGE] Enable upstream deployments

### PR DESCRIPTION
**Pull request by mfojtik**
Just a placeholder for the initial work of enabling upstream deployments in origin.

- [x] Enable upstream controller
- [x] Update SA policies to allow controller creating/updating RS
- [ ] Write conversions from D <-> DC
- [ ] Add non-convertible fields as an annotation
- [ ] Add "original-kind" annotation to allow controllers only to serve types they own
- [ ] Cohabitate `Deployment` and `DeploymentConfig`
- [ ] Tests
- [ ] Check the UI impact (does the D show in the UI?, could they be scaled?, etc.)

FYI: This is based on @kargakis PR, it will be squashed once his PR lands.

More info: https://github.com/openshift/origin/pull/9941
ReplicaSet: https://github.com/openshift/origin/pull/8274

### CODE REVIEW

**Comment by mfojtik** [comment]
@deads2k @smarterclayton ignore the conversion code for now, still WIP. With this I can create upstream deployments, the replica set will be created and the pods will run. However, it seems like the cohabitation of those two is not working (`oc get dc/d` returns 404).

**Comment by mfojtik** [comment]
The cohabitation of rc<->rs seems not to work as well (checked the etcd, they are both stored separately)

File: vendor/k8s.io/kubernetes/pkg/api/v1/register.go [hunk 1]
```diff
@@ -97,6 +97,10 @@ func addKnownTypes(scheme *runtime.Scheme) {
 	scheme.AddKnownTypes(SchemeGroupVersion, &extensions.ReplicaSet{})
 	scheme.AddKnownTypes(SchemeGroupVersion, &extensions.ReplicaSetList{})
 
+	// Add deployments
+	scheme.AddKnownTypes(SchemeGroupVersion, &extensions.Deployment{})
```
Comment (mfojtik) [inline vendor/k8s.io/kubernetes/pkg/api/v1/register.go]: i'm not sure why is this needed (or if it is needed at all)

File: vendor/k8s.io/kubernetes/pkg/api/v1/register.go [hunk 2]
```diff
@@ -92,6 +93,10 @@ func addKnownTypes(scheme *runtime.Scheme) {
 	// Add common types
 	scheme.AddKnownTypes(SchemeGroupVersion, &unversioned.Status{})
 
+	// Add replica sets
+	scheme.AddKnownTypes(SchemeGroupVersion, &extensions.ReplicaSet{})
```
Comment (mfojtik) [inline vendor/k8s.io/kubernetes/pkg/api/v1/register.go]: @deads2k @smarterclayton do you know why this might be needed?

File: vendor/k8s.io/kubernetes/pkg/genericapiserver/storage_factory.go [hunk 3]
```diff
@@ -138,6 +138,25 @@ func (s *DefaultStorageFactory) AddCohabitatingResources(groupResources ...unver
 	}
 }
 
+func (s *DefaultStorageFactory) ForceStorageResource(groupResource, toForce unversioned.GroupResource) {
```
Comment (mfojtik) [inline vendor/k8s.io/kubernetes/pkg/genericapiserver/storage_factory.go]: @deads2k I guess we are getting rid of this patch

File: vendor/k8s.io/kubernetes/pkg/api/v1/register.go (hunk 2 above)
Comment (mfojtik) [inline vendor/k8s.io/kubernetes/pkg/api/v1/register.go]: I know now... we get panic if we don't register it.

Comment (deads2k) [inline vendor/k8s.io/kubernetes/pkg/api/v1/register.go]: > I know now... we get panic if we don't register it.

I just got as I was enabling.  I'm not sure this is the correct solution.

**Comment by mfojtik** [comment]
@smarterclayton @danmcp this is a must for the release.

**Comment by mfojtik** [comment]
Closed because of https://github.com/openshift/origin/pull/10046
